api/providers/tasks.py

```diff
@@ -99,6 +107,213 @@ def prepare_for_registration_bulk_creation(payload_hash, initiator_id, provider_
+        'Bulk creating draft registrations: [provider={}, schema={}, initiator={}, auto_approval={}]'.format(
+            provider._id,
+            schema._id,
+            initiator._id,
+            auto_approval
+        )
+    )
+
+    # Check and pick up registration rows for creation
+    registration_rows = RegistrationBulkUploadRow.objects.filter(upload__id=upload_id)
+    logger.info("Picked up [{}] registration rows for creation".format(len(registration_rows)))
+
+    count = 0
+    error_list = []
+    for row in registration_rows:
+        count += 1
+        logger.info("[{}, {}, {}, {}]".format(count, row.is_picked_up, row.is_completed, row.draft_registration))
+        if not dry_run:
+            is_successful = False
+            try:
+                is_successful = handle_registration_row(row, initiator, provider, schema,
+                                                        error_list, auto_approval=auto_approval)
+            except Exception as e:
+                logger.error(e)
+                logger.error(e.__cause__)
+            if not is_successful:
+                if row.draft_registration:
+                    row.draft_registration.delete()
+
+    logger.info(error_list)
+
+
+def handle_registration_row(row, initiator, provider, schema, error_list, auto_approval=False):
+
+    metadata = row.csv_parsed.get('metadata', {})
+    row_external_id = metadata.get('external id')
+    responses = row.csv_parsed.get('registration_responses', {})
+    auth = Auth(user=initiator)
+
+    # Node
+    node_id = metadata.get('project guid', '')
+    node = None
+    if node_id:
+        try:
+            node = AbstractNode.objects.get(guids___id=node_id, is_deleted=False, type='osf.node')
+        except AbstractNode.DoesNotExist:
+            logger.error('Nod does not exist: [node_id={}]'.format(node_id))
+            error_list.append(row_external_id)
+            return False
+        except AbstractNode.MultipleObjectsReturned:
+            logger.error('Multiple nodes returned: [node_id={}]'.format(node_id))
+            error_list.append(row_external_id)
+            return False
+
+    # Subject
+    subject_names = metadata.get('subjects', [])
+    all_subjects = []
+    for subject_name in subject_names:
+        subject_list = provider.all_subjects.filter(text=subject_name)
+        if not subject_list:
+            logger.error('Subject not found: [subject_name={}]'.format(subject_name))
+            error_list.append(row_external_id)
+            return False
+        if len(subject_list) > 1:
+            logger.error('Duplicate subjects found: [subject_name={}]'.format(subject_name))
+            error_list.append(row_external_id)
+            return False
+        subject = subject_list.first()
+        all_subjects.append(subject)
+        parent_subject = subject.parent
+        while parent_subject:
+            all_subjects.append(parent_subject)
+            parent_subject = parent_subject.parent
```

Comment:
I could be wrong, but I suspect you can call [`set_subjects_from_relationships()`](https://github.com/CenterForOpenScience/osf.io/blob/c0da58edb7dadd2f5872cee7f4bfbfb767a4175d/osf/models/mixins.py#L1140) on the draft and provider the function with a list of leaf subjects without including the parents, and the function will automatically find the parents and add them by calling [`expand_subject_hierarchy()`](https://github.com/CenterForOpenScience/osf.io/blob/c0da58edb7dadd2f5872cee7f4bfbfb767a4175d/osf/models/validators.py#L130).

Reply:
This helps. My plan was to use [set_subjects()](`https://github.com/CenterForOpenScience/osf.io/blob/c0da58edb7dadd2f5872cee7f4bfbfb767a4175d/osf/models/mixins.py#L1109`), but I think the one you suggested is the better one to use. Will do.

Reply:
Done and fixed